### ISSUE TITLE
Performance Improvement: Use date range instead of casting

### ISSUE THREAD

**Issue by parisholley** [metabase/metabase#4043]
On postgres (9.1), when using the date filter on question creation, selecting the current date results in the following where clause: 

cast(column as date) = cast(now() as date)

moving to the following instead results in a huge speed improvement

column >= date_trunc('day',now()) and column < date_trunc('day',now() + '1 day')

currently our read-only replica cries with the following error because the cast causes the query to run too long

```
ERROR:  canceling statement due to conflict with recovery
DETAIL:  User query might have needed to see row versions that must be removed.
```

⬇️  **Please click the 👍 reaction instead of leaving a `+1` or 👍  comment**

**Comment by parisholley** [comment]
issue seems to related to the fact that it is using a func call on date, vs. just using a between statement. looking through code for best way to handle it. a change would need to be made in m.q.expand to use a between, but maybe a bigger issue is there is no concept of time i think. so need to basicly do inclusive from start of day, until last second of day

**Comment by eyougo** [comment]
This issue seems that because the execute of the sql first query out some results for later processing, but at that moment, the results of the query was modified by the master db, the read-only slave found that the results was changed by row versions and could not be processed, so cancel the sql statement.

http://stackoverflow.com/questions/14592436/postgresql-error-canceling-statement-due-to-conflict-with-recovery

https://www.postgresql.org/docs/current/static/hot-standby.html#HOT-STANDBY-CONFLICT

**Comment by parisholley** [comment]
correct, the issue only exists because we are pointing at a replica. but if the query was more efficient, we'd also not run into the issue :) though when using custom sql, using the "with()" syntax to fetch data in a temporary place then query against it, is also another way to mitigate the issue

**Comment by eyougo** [comment]
Oh,  Let me try this way. Thanks for your advice.

**Comment by davidzhao** [comment]
We are seeing this too with redshift. It impacts performance by a few orders of magnitude there given how redshift utilizes the sort key.

Using cast, query times out after 5 minutes, as it scans through all the records to compare the date.
```
SELECT col1, col2, time FROM events
WHERE CAST(time as date)
   BETWEEN getdate() - INTERVAL '2 day' and getdate() - INTERVAL '1 day'
LIMIT 10
```
Explain:
```
XN Limit  (cost=0.00..40.00 rows=10 width=60)
  ->  XN Seq Scan on events  (cost=0.00..60138.66 rows=15035 width=60)
        Filter: (('2017-05-02 00:00:00'::timestamp without time zone <= date("time")) AND ('2017-05-03 00:00:00'::timestamp without time zone >= date("time")))
```

Directly compare, query returns in 5 seconds
```
SELECT col1, col2, time FROM events
WHERE time
   BETWEEN CURRENT_DATE - INTERVAL '2 day' and CURRENT_DATE - INTERVAL '1 day'
LIMIT 10
```
Explain:
```
XN Limit  (cost=0.00..0.15 rows=10 width=60)
  ->  XN Seq Scan on events  (cost=0.00..0.66 rows=44 width=60)
        Filter: (("time" >= '2017-05-02 00:00:00'::timestamp without time zone) AND ("time" <= '2017-05-03 00:00:00'::timestamp without time zone))
```

Notice the cost & estimated rows given by the DB

**Comment by rubenmch** [comment]
Happening to us with CrateDB. I think most databases would be affected by this, casting is very slow when the table has a large number of rows since it has to compute the function for every row in the table instead of just using an index search. 

We tested metabase with a small 1 million rows table and it took 30 seconds to complete. The same table in production has 60 million rows so it would take about half an hour to complete vs a few seconds without casting.

Example sql generated by metabase (current year filter, 1 million rows, takes **30 seconds** to complete with 100% cpu usage on a 6 core 39GB RAM cluster):

SELECT sum(("doc"."pruebaventaproducto"."productocantidad" * "doc"."pruebaventaproducto"."preciodeventaunitarioconimpuesto")) AS "sum", "doc"."pruebaventaproducto"."productonombre"
FROM "doc"."pruebaventaproducto"
WHERE **CAST(extract(year from "doc"."pruebaventaproducto"."ventafecha") AS integer) = CAST(extract(year from current_timestamp(3)) AS integer)**
GROUP BY "doc"."pruebaventaproducto"."productonombre"
ORDER BY "sum" DESC, "doc"."pruebaventaproducto"."productonombre" ASC
LIMIT 25

Same query without date casting (takes **0.1 second** to complete):

SELECT sum(("doc"."pruebaventaproducto"."productocantidad" * "doc"."pruebaventaproducto"."preciodeventaunitarioconimpuesto")) AS "sum", "doc"."pruebaventaproducto"."productonombre"
FROM "doc"."pruebaventaproducto"
WHERE **"doc"."pruebaventaproducto"."ventafecha" > date_format('%Y-01-01', current_timestamp(3))**
GROUP BY "doc"."pruebaventaproducto"."productonombre"
ORDER BY "sum" DESC, "doc"."pruebaventaproducto"."productonombre" ASC
LIMIT 25;

**Comment by majnu-campact** [comment]
+1 Queries should use the underlying datatype so that no additional indexing is necessary (wherever possible). 

We created an additional postgres index on `(CAST(some_timestamp AS date))` just to make some filters perform in decent time. I understand that casting makes the SQL look prettier, but having to create additional indexes is a bit of a pain and waste of resources.

**Comment by lp-lima** [comment]
Same happening to me in MySQL. Apparently, MySql doesn't allow the creation of function based index. IT means that I have no workaround to handle this issue.

**Comment by wmorin** [comment]
I can confirm this issue on Redshift too.

**Comment by GabrielCampbellHaud** [comment]
Same problem with V0.29.3 was wondering if any1 has some update

**Comment by paulverdier** [comment]
I'm having the same issue. Would like to see a fix.

**Comment by wmorin** [comment]
Metabase would be super fast on Redshift with this change 🚀.

**Comment by JoeSham** [comment]
Also would appreciate this issue. For _timestamp_ columns it can be solved with adding an index like:
```
create index table_ts_as_date_idx on booking_all_v2 (cast(ts as date));
```
But for _timestamp with time zone_ column we can't even create the index as it is not immutable. (Or maybe we just haven't found how yet.)

**Comment by lp-lima** [comment]
Not even that is possible in MySQL 5.7 - therefore, we here have to endure a painful full table scan to retrieve data we could get in literally 1 second using SQL

**Comment by JoeSham** [comment]
@salsakran Hi, can you please give update if / when do you plan on solving this issue? Thx.

**Comment by wmorin** [comment]
Super interested in this feature too!

**Comment by Nyanna** [comment]
Hi,

we use a Postgres int8 to store a ms epoch timestamp.
Our workaround was to use a materialized view instead of a mutable function index to retain the click adventure metabase provides to our product managers and use indices to speedup queries. 

```sql
create materialized view surveys_tsd as
select uid, surveynr, answerid, freetext, cid, (CAST(to_timestamp(ts / 1000) as date)) as date from surveys;
```

for metabased build query of:

```sql
explain verbose SELECT CAST("public"."surveys_tsd"."date" AS date) AS "date", count(*) AS "count"
FROM "public"."surveys_tsd"
WHERE CAST("public"."surveys_tsd"."date" AS date) BETWEEN CAST((NOW() + INTERVAL '-30 day') AS date)
   AND CAST((NOW() + INTERVAL '-1 day') AS date)
GROUP BY CAST("public"."surveys_tsd"."date" AS date)
ORDER BY CAST("public"."surveys_tsd"."date" AS date) asc;
```

```sql
GroupAggregate  (cost=0.45..3587.87 rows=25 width=4)
  Output: date, count(*)
  Group Key: surveys_tsd.date
  ->  Index Only Scan using surveys_tsd_date_idx on public.surveys_tsd  (cost=0.45..3204.85 rows=76554 width=4)
        Output: date
        Index Cond: ((surveys_tsd.date >= ((now() + '-30 days'::interval))::date) AND (surveys_tsd.date <= ((now() + '-1 days'::interval))::date))
```

**Comment by lp-lima** [comment]
I just noticed that, using 0.31, I'm not seeing this issue anymore. Since metabase is using CONVERT_TZ for MySQL, it is not doing the date() function over the columns anymore. Therefore, I withdraw my request from here, since, for me, the issue is already solved.

**Comment by rhnBen** [comment]
Using intervals instead of casting may cause problems when you are dealing with "daylight saving" concept.
In these cases, the difference between two dates may be 23 hours or 25 hours but '1 day' interval is equal to 24 hours. If you want to using intervals, make sure you're handling these cases correctly.

**Comment by xqwbx163** [comment]
> I just noticed that, using 0.31, I'm not seeing this issue anymore. Since metabase is using CONVERT_TZ for MySQL, it is not doing the date() function over the columns anymore. Therefore, I withdraw my request from here, since, for me, the issue is already solved.

but i found this issue add the date() function on metabase 0.31,can you help me?

**Comment by squeaky-pl** [comment]
This also breaks effective access to range partitioned tables on Postgres, and I can imagine other database that support partitioning.

We have `metrics` table partitioned  into `metric_<year>_<month>` partitions.

A well-constructed query for one day of data will first eliminate all the unneeded partitions and then scan an index to read few megabytes of data:

```
EXPLAIN SELECT FROM metrics WHERE timestamp BETWEEN '2019-01-18' AND '2019-01-19';

                                                                               QUERY PLAN                                                                                
-------------------------------------------------------------------------------------------------------------------------------------------------------------------------
 Append  (cost=12.03..25824.00 rows=1 width=0)
   ->  Bitmap Heap Scan on metrics_2019_01  (cost=12.03..25824.00 rows=1 width=0)
         Recheck Cond: (("timestamp" >= '2019-01-18 00:00:00+00'::timestamp with time zone) AND ("timestamp" <= '2019-01-19 00:00:00+00'::timestamp with time zone))
         ->  Bitmap Index Scan on ix_metrics_2019_01_timestamp  (cost=0.00..12.03 rows=11148 width=0)
               Index Cond: (("timestamp" >= '2019-01-18 00:00:00+00'::timestamp with time zone) AND ("timestamp" <= '2019-01-19 00:00:00+00'::timestamp with time zone))
``` 

but as soon as you use an expression on timestamp (which Metabase does) you will read all the partitions and then sequentially scan them one by, rendering both partition elimination and index useless. This reads hundreds of GBs from disk.

```
sling_analytics=> EXPLAIN SELECT FROM metrics WHERE DATE(timestamp) BETWEEN '2019-01-18' AND '2019-01-19';
                                                   QUERY PLAN                                                    
-----------------------------------------------------------------------------------------------------------------
 Gather  (cost=1000.00..126091.63 rows=30843 width=0)
   Workers Planned: 2
   ->  Append  (cost=0.00..122007.33 rows=12851 width=0)
         ->  Parallel Seq Scan on metrics_2018_12  (cost=0.00..44766.82 rows=4720 width=0)
               Filter: ((date("timestamp") >= '2019-01-18'::date) AND (date("timestamp") <= '2019-01-19'::date))
         ->  Parallel Seq Scan on metrics_2019_01  (cost=0.00..77225.21 rows=8130 width=0)
               Filter: ((date("timestamp") >= '2019-01-18'::date) AND (date("timestamp") <= '2019-01-19'::date))
         ->  Parallel Seq Scan on metrics_2019_02  (cost=0.00..15.29 rows=1 width=0)
               Filter: ((date("timestamp") >= '2019-01-18'::date) AND (date("timestamp") <= '2019-01-19'::date))
(9 rows)

```

This is a very common way of organizing timeseries data and accessing it effectively by time ranges.

**Comment by Luckymethod** [comment]
I join the chorus of folks that would like a resolution for this.

**Comment by parisholley** [comment]
With my limited clojure experience, digging through the code base, i found two lines of interest.

Most of the drivers are piping date expressions to honeysql which I assume is forcing the cast to happen.

https://github.com/metabase/metabase/blob/6fd6dbc73c8a3f44691d0eb9ae1165b3a86c2d2b/src/metabase/driver/postgres.clj#L123

Additionally, it would appear that we intentionally detect if a field is a date, and run it through the driver vs. just passing the field down without translating. It feels like this was trying to solve a specific use case (ie: casting left side to match right side in group by, etc), but I don't have that historical context.

https://github.com/metabase/metabase/blob/a251fbf9d4f52b8df2499dd0e8b295bd66f7c2d5/src/metabase/query_processor/middleware/parameters/sql.clj#L370

My vote would be to change the code to ONLY run the field mapping for date fields through driver if a certain flag/parameter is set at runtime (perhaps, based on the type of query?), vs. assuming we should do it in all cases.

**Comment by nikita2206** [comment]
Hey, I've made a relatively dirty fix for this (dirty in the sense of code quality/architecture, not that it breaks anything). If you need these optimizations you can use my branch from #9860 PR to build your own Docker image with them included. Hopefully I'll come up with something that we will be okay to merge later during the month.

**Comment by kmatarese** [comment]
Metabase was a leading candidate for our next analytics solution until I hit this (v0.32.9 with MySQL back end). Unfortunately it makes the product effectively unusable for us. This seems like a critical performance bug for at least a few major databases. I'm surprised it's taken over two years to address, but also curious what people are doing differently that avoids the hit in case I'm just missing something.

**Comment by lp-lima** [comment]
> Metabase was a leading candidate for our next analytics solution until I hit this (v0.32.9 with MySQL back end). Unfortunately it makes the product effectively unusable for us. This seems like a critical performance bug for at least a few major databases. I'm surprised it's taken over two years to address, but also curious what people are doing differently that avoids the hit in case I'm just missing something.

For us here, we use MySQL, and, when I query things using a "BETWEEN" filter, metabase does not cast, but rather uses the column itself (as of 0.31.0, which is my current metabase version).

**Comment by nikita2206** [comment]
@lp-lima you must be querying by specific datetimes instead of dates

@kmatarese the above could somewhat help (it should help but it's just not convenient to do), i.e. when using this filter use absolute dates, not relative ones, and always specify date _and_ time

**Comment by kmatarese** [comment]
Thanks for the suggestions guys. I wouldn't be able to enforce that workaround with our team, eventually someone will forget to do it and kick off very expensive queries. I'll check back in down the road in case this gets addressed before we go another direction.

**Comment by lp-lima** [comment]
> @lp-lima you must be querying by specific datetimes instead of dates
> 
> @kmatarese the above could somewhat help (it should help but it's just not convenient to do), i.e. when using this filter use absolute dates, not relative ones, and always specify date _and_ time

I didn't quite understand what were you refering to. I mean, I said that, when using between, metabase behaves properly, but not otherwise. There are no datetimes involved at all, at any point, only dates...

**Comment by jdsterritt** [comment]
I'm new to Metabase and have been impressed with the product thus far. I am testing the product as a candidate for our analytics solution, but as mentioned above this issue essentially makes the product unusable.

**Comment by camsaul** [comment]
Hey everyone, this is definitely something we plan on fixing in the near future. Stay tuned!

**Comment by nikita2206** [comment]
@camsaul just to remind that unless you’re going to go with a different approach/implementation, there’s a PR which fixes performance and it currently is blocked by https://github.com/metabase/mbql/pull/19

**Comment by klyhne** [comment]
I am experiencing the issue with date type columns are being cast to timestamp when summarizing on Date[Month].
We are running Metabase 0.35.1 on a PosgreSQL database version 11.1

The query I'm running is:
![Metabase_date_cast](https://user-images.githubusercontent.com/10856795/87151576-66a1c100-c2b4-11ea-88b8-346d0b0498ad.png)

And the resulting SQL is:
```
SELECT date_trunc('month', CAST("dw"."campaign_events"."date" AS timestamp)) AS "date", count(*) AS "count"
FROM "dw"."campaign_events"
GROUP BY date_trunc('month', CAST("dw"."campaign_events"."date" AS timestamp))
ORDER BY date_trunc('month', CAST("dw"."campaign_events"."date" AS timestamp)) ASC
```

The data model for the table in Metabase looks like this:
![Metabase_data_model](https://user-images.githubusercontent.com/10856795/87152017-327ad000-c2b5-11ea-827f-e89b834b829c.png)

The PostgreSQL view that the table select from looks like this:
![PostgreSQL_view](https://user-images.githubusercontent.com/10856795/87152043-40c8ec00-c2b5-11ea-9872-f8a1ab4b4cdc.png)

If I aggregate on Timestamp[Month] istead of Date[Month], the result is the same. The resulting SQL will CAST the timestamp column to type timestamp, resulting in a very slow query.

Below is the diagnostics info from Metabase:
```
{
  "browser-info": {
    "language": "en-US",
    "platform": "Win32",
    "userAgent": "Mozilla/5.0 (Windows NT 10.0; Win64; x64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/83.0.4103.116 Safari/537.36",
    "vendor": "Google Inc."
  },
  "system-info": {
    "file.encoding": "UTF-8",
    "java.runtime.name": "OpenJDK Runtime Environment",
    "java.runtime.version": "1.8.0_242-8u242-b08-0ubuntu3~16.04-b08",
    "java.vendor": "Private Build",
    "java.vendor.url": "http://java.oracle.com/",
    "java.version": "1.8.0_242",
    "java.vm.name": "OpenJDK 64-Bit Server VM",
    "java.vm.version": "25.242-b08",
    "os.name": "Linux",
    "os.version": "4.4.0-138-generic",
    "user.language": "en",
    "user.timezone": "Etc/UTC"
  },
  "metabase-info": {
    "databases": [
      "postgres"
    ],
    "hosting-env": "unknown",
    "application-database": "h2",
    "application-database-details": {
      "database": {
        "name": "H2",
        "version": "1.4.197 (2018-03-18)"
      },
      "jdbc-driver": {
        "name": "H2 JDBC Driver",
        "version": "1.4.197 (2018-03-18)"
      }
    },
    "run-mode": "prod",
    "version": {
      "tag": "v0.35.1",
      "date": "2020-04-02",
      "branch": "release-0.35.x",
      "hash": "e67f169"
    },
    "settings": {
      "report-timezone": "Europe/Berlin"
    }
  }
}
```

Any help would be much appreciated! I don't know if the issue should be reopened.

**Comment by flamber** [comment]
@klyhne This was specific to the filters. There's another issue, which focuses on the double-casting (which I originally thought was a duplicate of this) - #9934 - upvote by clicking :+1: on the first post

**Comment by laratutoteam** [comment]
Easy Solution : https://laratuto.com/date-between-query-in-mysql/
